File: rqa.py
import math
import random
import time

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def __repr__(self):
        current = self.head
        nodes = []
        while current:
            nodes.append(current.data)
            current = current.next
        return ' -> '.join(map(str, nodes))

def read_points_from_file(filename):
    points = []
    with open(filename, 'r') as file:
        for line in file:
            coordinates = list(map(float, line.strip().split(' ')))
            points.append(tuple(coordinates))
    return points

def append_query_points(query_filename, normalized_points):
    query_points = read_points_from_file(query_filename)
    start_index = len(normalized_points)  
    normalized_points.extend(query_points)
    return list(range(start_index, len(normalized_points)))  

def choose_reference_point(points):
    return random.choice(points)

def distance(a, b):
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(a, b)))

def radius_query(points, radius, reference, query_indices):
    dist = []
    length_X = len(points)
    for i in range(length_X):
        d = round(distance(reference, points[i]), 1)
        dist.append(d)
    
    sorted_X_indices = [i for i, _ in sorted(enumerate(dist), key=lambda x: x[1])]
    sorted_X = [points[i] for i in sorted_X_indices]

    hash_table = {}

    for i in sorted_X_indices:
        hash_table[i] = LinkedList()

    for i in range(length_X):
        for j in range(i + 1, length_X):
            if distance(reference, sorted_X[j]) - distance(reference, sorted_X[i]) > radius:
                break
            if distance(sorted_X[i], sorted_X[j]) <= radius:
                hash_table[sorted_X_indices[i]].append(sorted_X_indices[j])
                hash_table[sorted_X_indices[j]].append(sorted_X_indices[i])

    return {key: hash_table[key] for key in query_indices}


def write_result_to_file(hash_table, filename):
    with open(filename, 'w') as file:
        file.write("===============================================================\n")
        i=0
        for key, value in hash_table.items():
            file.write(f"Result[{i}]: {value}\n") 
            i+=1

if __name__ == "__main__":
    normalized_points_filename = "lsh_normalized_points.txt"
    query_filename = "lsh_query_points.txt"
    radius = 0.6
    
    normalized_points = read_points_from_file(normalized_points_filename)
    query_points = read_points_from_file(query_filename)
    print(f"The total number of Given data points are: {len(normalized_points)}")
    print(f"The total number of Query Points are: {len(query_points)}")
    print(f"The dimensions of the points is: {len(normalized_points[0])}")
    print(f"The given radius is: {radius}")
    query_indices = append_query_points(query_filename, normalized_points)
    print(f"The total number of processing points after appending query points are: {len(normalized_points)}")

    reference = choose_reference_point(normalized_points)
    
    start_time = time.time()
    hash_table = radius_query(normalized_points, radius, reference, query_indices)
    end_time = time.time()

    result_filename = "result.txt"
    write_result_to_file(hash_table, result_filename)
    print(f"Result is available in \"{result_filename}\" file.")
    print("========================================================")
    print(f"The time taken is {end_time - start_time} seconds.")

File: rqa.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <list>
#include <unordered_map>
#include <random>
#include <ctime>
#include <algorithm>
#include <sstream>

struct Node {
    int data;
    Node* next;
    Node(int data) : data(data), next(nullptr) {}
};

class LinkedList {
public:
    Node* head;
    LinkedList() : head(nullptr) {}

    void append(int data) {
        Node* new_node = new Node(data);
        if (!head) {
            head = new_node;
        } else {
            Node* current = head;
            while (current->next) {
                current = current->next;
            }
            current->next = new_node;
        }
    }

    friend std::ostream& operator<<(std::ostream& os, const LinkedList& list) {
        Node* current = list.head;
        while (current) {
            os << current->data << " -> ";
            current = current->next;
        }
        return os << "NULL";
    }
};

std::vector<std::vector<double>> readPointsFromFile(const std::string& filename) {
    std::vector<std::vector<double>> points;
    std::ifstream file(filename);
    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::vector<double> point;
        double coord;
        while (iss >> coord) {
            point.push_back(coord);
        }
        points.push_back(point);
    }
    return points;
}

std::vector<double> chooseReferencePoint(const std::vector<std::vector<double>>& points) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(0, points.size() - 1);
    return points[distrib(gen)];
}

double distance(const std::vector<double>& a, const std::vector<double>& b) {
    double dist = 0.0;
    for (size_t i = 0; i < a.size(); ++i) {
        dist += (a[i] - b[i]) * (a[i] - b[i]);
    }
    return std::sqrt(dist);
}

std::unordered_map<int, LinkedList> radiusQuery(const std::vector<std::vector<double>>& points, double radius, const std::vector<double>& reference, const std::vector<int>& queryIndices) {
    std::vector<std::pair<double, int>> dist(points.size());
    for (size_t i = 0; i < points.size(); ++i) {
        dist[i] = {distance(reference, points[i]), i};
    }
    std::sort(dist.begin(), dist.end());

    std::unordered_map<int, LinkedList> hashTable;
    for (auto& p : dist) {
        hashTable[p.second] = LinkedList();
    }

    for (size_t i = 0; i < dist.size(); ++i) {
        for (size_t j = i + 1; j < dist.size() && (dist[j].first - dist[i].first) <= radius; ++j) {
            if (distance(points[dist[i].second], points[dist[j].second]) <= radius) {
                hashTable[dist[i].second].append(dist[j].second);
                hashTable[dist[j].second].append(dist[i].second);
            }
        }
    }

    std::unordered_map<int, LinkedList> result;
    for (int idx : queryIndices) {
        result[idx] = hashTable[idx];
    }
    return result;
}

void writeResultToFile(const std::unordered_map<int, LinkedList>& hashTable, const std::string& filename) {
    std::ofstream file(filename);
    file << "===============================================================\n";
    std::string queryFilename = "lsh_query_points.txt";
    auto queryPoints = readPointsFromFile(queryFilename);
    int i = queryPoints.size();
    for (auto& pair : hashTable) {
        file << "Result[" << i << "]: " << pair.second << "\n";
        i--;
    }
}

int main() {
    std::string normalizedPointsFilename = "lsh_normalized_points.txt";
    std::string queryFilename = "lsh_query_points.txt";
    double radius = 0.6;

    auto normalizedPoints = readPointsFromFile(normalizedPointsFilename);
    auto queryPoints = readPointsFromFile(queryFilename);

    std::vector<int> queryIndices;
    for (size_t i = normalizedPoints.size(); i < normalizedPoints.size() + queryPoints.size(); ++i) {
        queryIndices.push_back(i);
    }
    normalizedPoints.insert(normalizedPoints.end(), queryPoints.begin(), queryPoints.end());

    auto reference = chooseReferencePoint(normalizedPoints);
    
    clock_t start = clock();
    auto hashTable = radiusQuery(normalizedPoints, radius, reference, queryIndices);
    clock_t end = clock();

    std::string resultFilename = "result.txt";
    writeResultToFile(hashTable, resultFilename);

    double elapsed_time = static_cast<double>(end - start) / CLOCKS_PER_SEC;
    std::cout << "Result is available in \"" << resultFilename << "\" file.\n";
    std::cout << "========================================================\n";
    std::cout << "Time taken: " << elapsed_time << " seconds.\n";

    return 0;
}